File: frontend/src/services/authService.js
import { API_ENDPOINTS, getAuthHeaders, getAccessToken, getRefreshToken } from './config';

class AuthService {
  constructor() {
    this.baseURL = API_ENDPOINTS.AUTH;
  }

  // Almacenar tokens en localStorage
  setTokens(accessToken, refreshToken) {
    if (accessToken) localStorage.setItem('access_token', accessToken);
    if (refreshToken) localStorage.setItem('refresh_token', refreshToken);
  }

  // Limpiar tokens de localStorage
  clearTokens() {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('user');
  }

  // Wrappers para obtener tokens (usa las funciones de config si están disponibles)
  getAccessToken() {
    try {
      return getAccessToken();
    } catch {
      return localStorage.getItem('access_token');
    }
  }

  getRefreshToken() {
    try {
      return getRefreshToken();
    } catch {
      return localStorage.getItem('refresh_token');
    }
  }

  // Registro de usuario
  async register(userData) {
    try {
      const response = await fetch(`${this.baseURL}/register/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(userData)
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || data.error || 'Error en el registro');
      }

      // Almacenar tokens si el registro incluye login automático
      if (data.tokens) {
        this.setTokens(data.tokens.access, data.tokens.refresh);
        if (data.user) {
          localStorage.setItem('user', JSON.stringify(data.user));
        }
      }

      return {
        success: true,
        data: data,
        user: data.user,
        tokens: data.tokens
      };
    } catch (error) {
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Login de usuario
  async login(credentials) {
    try {
      const response = await fetch(`${this.baseURL}/login/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(credentials)
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || data.error || 'Credenciales inválidas');
      }

      // Almacenar tokens y datos del usuario
      if (data.tokens) {
        this.setTokens(data.tokens.access, data.tokens.refresh);
      }
      
      if (data.user) {
        localStorage.setItem('user', JSON.stringify(data.user));
      }

      return {
        success: true,
        data: data,
        user: data.user,
        tokens: data.tokens
      };
    } catch (error) {
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Logout de usuario
  async logout() {
    try {
      const refreshToken = getRefreshToken();
      
      if (refreshToken) {
        await fetch(`${this.baseURL}/logout/`, {
          method: 'POST',
          headers: getAuthHeaders(true),
          body: JSON.stringify({ refresh: refreshToken })
        });
      }
    } catch (error) {
      console.error('Error durante logout:', error);
    } finally {
      // Limpiar tokens independientemente del resultado
      this.clearTokens();
    }

    return { success: true };
  }

  // Obtener perfil del usuario autenticado
  async getProfile() {
    try {
      const response = await fetch(`${this.baseURL}/profile/`, {
        method: 'GET',
        headers: getAuthHeaders(true)
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || data.error || 'Error al obtener perfil');
      }

      return {
        success: true,
        data: data
      };
    } catch (error) {
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Refrescar token de acceso
  async refreshAccessToken() {
    try {
      const refreshToken = getRefreshToken();
      
      if (!refreshToken) {
        throw new Error('No hay refresh token disponible');
      }

      const response = await fetch(`${this.baseURL}/token/refresh/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ refresh: refreshToken })
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error('Token refresh failed');
      }

      // Actualizar access token
      localStorage.setItem('access_token', data.access);

      return {
        success: true,
        accessToken: data.access
      };
    } catch (error) {
      // Si falla el refresh, limpiar tokens
      this.clearTokens();
      return {
        success: false,
        error: error.message
      };
    }
  }

  // Obtener usuario desde localStorage
  getCurrentUser() {
    try {
      const userStr = localStorage.getItem('user');
      return userStr ? JSON.parse(userStr) : null;
    } catch {
      return null;
    }
  }
}

// Exportar instancia singleton
export const authService = new AuthService();
export default authService;

File: mobile_app/lib/services/auth_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/user_model.dart';

/// Servicio de Autenticación
class AuthService {
  static const String baseUrl = 'http://localhost:8000/api/auth';

  // Claves para almacenamiento local
  static const String _accessTokenKey = 'access_token';
  static const String _refreshTokenKey = 'refresh_token';
  static const String _userDataKey = 'user_data';

  // === REGISTRO ===

  Future<Map<String, dynamic>> register({
    required String email,
    required String username,
    required String password,
    required String firstName,
    required String lastName,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/register/'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'email': email,
        'username': username,
        'password': password,
        'first_name': firstName,
        'last_name': lastName,
      }),
    );

    if (response.statusCode == 201) {
      final data = jsonDecode(response.body);
      await _saveTokens(data['tokens']['access'], data['tokens']['refresh']);
      await _saveUserData(data['user']);
      return data;
    } else {
      throw Exception('Error en registro: ${response.body}');
    }
  }

  // === LOGIN ===

  Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/login/'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      await _saveTokens(data['tokens']['access'], data['tokens']['refresh']);
      await _saveUserData(data['user']);
      return data;
    } else {
      throw Exception('Error en login: ${response.body}');
    }
  }

  // === LOGOUT ===

  Future<void> logout() async {
    final refreshToken = await getRefreshToken();

    if (refreshToken != null) {
      try {
        await http.post(
          Uri.parse('$baseUrl/logout/'),
          headers: {'Content-Type': 'application/json'},
          body: jsonEncode({'refresh': refreshToken}),
        );
      } catch (e) {
        // Ignorar errores de logout en el servidor
      }
    }

    await _clearTokens();
  }

  // === GESTIÓN DE TOKENS ===

  Future<void> _saveTokens(String accessToken, String refreshToken) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_accessTokenKey, accessToken);
    await prefs.setString(_refreshTokenKey, refreshToken);
  }

  Future<String?> getAccessToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_accessTokenKey);
  }

  Future<String?> getRefreshToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_refreshTokenKey);
  }

  Future<void> _clearTokens() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_accessTokenKey);
    await prefs.remove(_refreshTokenKey);
    await prefs.remove(_userDataKey);
  }

  // === GESTIÓN DE USUARIO ===

  Future<void> _saveUserData(Map<String, dynamic> userData) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_userDataKey, jsonEncode(userData));
  }

  Future<User?> getCurrentUser() async {
    final prefs = await SharedPreferences.getInstance();
    final userDataString = prefs.getString(_userDataKey);

    if (userDataString != null) {
      final userData = jsonDecode(userDataString);
      return User.fromJson(userData);
    }

    return null;
  }

  Future<bool> isAuthenticated() async {
    final token = await getAccessToken();
    return token != null;
  }

  // === PERFIL ===

  Future<Map<String, dynamic>> getProfile() async {
    final token = await getAccessToken();

    if (token == null) {
      throw Exception('Usuario no autenticado');
    }

    final response = await http.get(
      Uri.parse('$baseUrl/profile/'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Error al obtener perfil: ${response.body}');
    }
  }
}
